### ISSUE TITLE
Basic Auth with external authentication service

### ISSUE THREAD

**Issue by giohappy** [geosolutions-it/geoserver#145]
I would like to discuss the opportunity to implement Basic Auth authentication through an external service. 

The idea is to have something similar to the ["Key provider using an external service"] (https://docs.geoserver.org/stable/en/user/community/authkey/index.html#key-provider-using-an-external-web-service) offered by the Key authentication module.

I don't know the details of the Basic Auth implementation in Geoserver, so I'm not able to suggest if this could / should be provided as a configurable option within the core Basic Auth, or if a new extension is required.
We might also opt to extend the Key auth module itself, forcing its original intents. 

I would like to hear the opinions from @aaime and the others from the Geoserver team, since I will need to estimate the effort soon.

**Comment by simboss** [comment]
@giohappy which geoserver version are we targeting?

**Comment by aaime** [comment]
I have no idea what "Basic Auth authentication through an external service" means? :-D

**Comment by simboss** [comment]
@nmco can you groom this with @aaime help and then assign it to someone

**Comment by giohappy** [comment]
@aaime here it is our use case.

The AuthKey module permits to authenticate a user by parsing an authkey parameter. Then my intention is to map that user to its role by means of a an AuthKey REST Roles service.

I would like to implement something similar for Basic Auth access. From my understanding Geoserver's Basic Auth filter only extracts username and password credentials from the request, then it delegates the authentication to the Authentication Providers. The available AP are "username and password" (which uses a UserGroups service to perform the authentication), "LDAP" and "JDBC".
If we had a "Web Service Authentication Provider", able to perform the authentication of the user credentials (obtained from the Basic Auth filter) and return the username, I should have (in few workds) "Basic Auth with external authentication service"

If I've misunderstood the Geoserver security concepts please correct me.
I'm also assuming that retrieving the username from the Web Service (either through the ApiKey Module or this new AP that I'm proposing) is enough, and the Roles Service will do the rest. Am I wrong?

**Comment by aaime** [comment]
@giohappy ok now it makes sense (well some, I don't know exactly how authkey works, I wrote it originally but it went through 2 full rewrites and does not work at all like it used to). 
I have looked into the code to figure out a bit how it works, here is what I've found, best if @mbarto and/or @afabiani confirm it though, I know very little about the GeoServer authentication part (the area that I know more is authorization):

- The authentication filter (how one providers the credentials and validates them): e.g., [GeoServerBasicAuthenticationFilter](https://github.com/geoserver/geoserver/blob/8b41abb4294e94c8fa10e6793f72a466a70d62e3/src/main/src/main/java/org/geoserver/security/filter/GeoServerBasicAuthenticationFilter.java#L33) which in turn uses a local user catalog to compare user name and password against. In authkey that can be delegated to an external service through [WebServiceAuthenticationKeyMapper](https://github.com/geoserver/geoserver/blob/e893e997434f0ca1b085896e27b1d774f8cafb6a/src/extension/authkey/src/main/java/org/geoserver/security/WebServiceAuthenticationKeyMapper.java#L40) which talks to an external service in order to verify an authentication
- After successful authentication, a way to associate roles past successful authentication, which is done by a GeoServerRoleService, e.g. [GeoServerRestRoleService](https://github.com/geoserver/geoserver/blob/e893e997434f0ca1b085896e27b1d774f8cafb6a/src/extension/authkey/src/main/java/org/geoserver/security/GeoServerRestRoleService.java)

As far as I can see those end up doing two separate remote requests, and there [is a caching mechanism](https://github.com/geoserver/geoserver/blob/ad7eb9041ff5dab135cba61798e8415c3b91097d/src/extension/authkey/src/main/java/org/geoserver/security/GeoServerAuthenticationKeyFilter.java#L83) in the middle to avoid doing too many requests.

So if I understand correctly, you'd like to extract username and password from the basic auth, and then do remote calls to verify identity and grab roles. I don't know if it can be done with a single call, authkey does it with two following 1-1 the Spring Security architecture (which GeoServer uses with some customization). And I'm guessing the remote calls won't be the same expected by authkey?

**Comment by giohappy** [comment]
> After successful authentication, a way to associate roles past successful authentication, which is done by a GeoServerRoleService, e.g. GeoServerRestRoleService

At least this is what I expect to happen, but I still have to veirfy it this is expected to work and it actually works. I have both the WebServiceAuthenticationKeyMapper which returns:

`{"username":"giovanni"}
`

 and the GeoServerRestRoleService which returns:

```
#/api/users:
{
  "users": [
    {
      "username": "giovanni",
      "groups": [
        "free"
      ]
    }
  ]
}
```

and 

```
#/api/roles:
{
  "groups": [
    "pro",
    "enterprise",
    "hub",
    "free",
    "admin"
  ]
}
```

Roles appear correctly inside Geoserver roles tab but but I'm not able to authenticate "giovanni" yet. If my assumptions are correct either I have something wrong with thre regex or something else is not working :)

In any case, yes, I want something similar to verify username and password credentials:

 - verify user credentials -> get the username
 - map the username to a role

Also geonode-allauth does two calls, afaik, one to validate the access_token key and retrieve the user infos, and then GeoServerRestRoleService to map him to the roles.

That's why I was thinking to adopt the same pattern here.

**Comment by giohappy** [comment]
Actually it seems that the same AuthKey WebService can be used as a mix of a UserGroup service and a Role service. According to the paragraph "Optional static comma-separated list of available Groups from the Web Service response" under [this](https://docs.geoserver.org/stable/en/user/community/authkey/index.html#authkey-webservice-body-response-usergroup-service) it states the following:

> It is worth notice that this UserGroup Service will always translate fetched Roles in the form ROLE_<ROLENAME>

from this sentence it seems that an "AuthKEY WebService Body Response" GroupService could be used to retrieve user roles...
@mbarto please shed some light on this :)

**Comment by giohappy** [comment]
> So if I understand correctly, you'd like to extract username and password from the basic auth, and then do remote calls to verify identity and grab roles. I don't know if it can be done with a single call, authkey does it with two following 1-1 the Spring Security architecture (which GeoServer uses with some customization). And I'm guessing the remote calls won't be the same expected by authkey?

I don't know the internal details of AuthKey, but yes both AuthKey and geonode-oauth2 (which are the two modules we currently use) do two calls. First one authenticates the user and retrieves its rolenames and the second one binds them through their respective roles services.

Roughly this is what happens (from our tests)

AuthKey:
 - AuthKey Filter -> retrieves users from the WebService Body response (through an apikey token)
 - AuthKey UserGroup webservice -> retrieves user's roles from the WebService Body response
 - AuthKey UserGroup webservice -> binds user's roles to GrantedAuthorities through ApiKey RESTO Roles service

geonode-oauth2:
 - geonode-oauth2 Auth Filter -> retrieves user from an Oauth2 endpoint (through an access token)
 - geonode-oauth2 -> retrieved and binds user's roles to GrantedAuthorities through ApiKey RESTO Roles service
In the end it seems that these modules shortcut the configuration of Filters + Providers. 
I can't say what's the right / best approach, considering the underlying Spring framework and the 
Geoserver's security model. 

Options:

 - extend the Baisc Auth Filter to authenticate through an external WebService + Roles Service, similar to what AuthKey and geonode-oauth2 do
 - extend the username/password authentication provider to support external Web Service + Roles Service

**Comment by nmco** [comment]
@imranrajjad before you move on with this one wait for my initial investigation.

**Comment by imranrajjad** [comment]
@nmco understood

**Comment by nmco** [comment]
Unfortunately I will have not time this or next week to perform the initial investigation, hence I will ask your help whit this @imranrajjad.

Please have look at Giovanni and Andrea feedback and then let's have a quick chat.

**Comment by simboss** [comment]
usual thing @imranrajjad:

- check the requirements, makre sure they are clear, ask questions if they are not
- check code in GS, propose a solution that align with GS codebase
- provide an estimate and wait for the GO

**Comment by imranrajjad** [comment]
Before sharing the plan and estimate, I am a bit unclear on the type of response which should be expected by the authenticating webservice. Is it a one time call which responds with roles already known to GS through available Role provider services? Or will it simply respond with a yes/no. 

If there is such a service available or atleast a specification is available, it will be really helpful.

### Plan

#### Web Authentication Provider
Extend `GeoServerAuthenticationProvider` and implement a functionality similar to `UsernamePasswordAuthenticationProvider`. User will be able to configure this provider with a URL, including place holders and response parsing scheme (e.g JSON path, Regex or Xpath)

#### Configuration
A configuration UI that will allow users to define external URL with required place holder
(e.g http://auth.service.com/auth?use={USER}&pass={passowrd}

![image](https://user-images.githubusercontent.com/11514936/74055447-e1af3480-4a01-11ea-8bc7-03cef2cc5dce.png)

User can also configure how to handle response. The response handling will differ with the response content types.

Initial estimate 3 days

**Comment by simboss** [comment]
@imranrajjad have read the full story above? Have you reviewed how authkey works?
If you did, well, do it again :).

In any case, you can ask details on the  service counterpart to @giohappy but, mind you, we are not developing something specific for DJAM but something rather generic like what we have for authkey.

Keep the discussion going here and keep @nmco in the loop.

**Comment by nmco** [comment]
Summarizing the actions points (for you) that resulted from our call @imranrajjad:
- Finish setting up AuthKey working with an external web service so you can properly debug how it works, ask @taba90 help, he just configure it last week for another issue.
- Investigate how AuthKey works, more precisely the integration points with Spring security to support the key external web service scenario. Describe your findings on this issue.
- Describe the possibilities you see to implemented the requested features along whit their _pos_ and _cons_:
  - A new authentication provider based on a web service and then a new roles provider?
  - A single call for the authentication and roles requesting?
  - Should we have a cache in between? 
- The implemented solution should be as much generic as possible, see for example how AuthKey supports a regex to get the roles from the web service response, hence creating a loose coupling between the client and the service.

Once we have a green light for one of the propose approaches, let's setup a detailed plan and review it, this will requrie a GSIP in GS :spades:.

**Comment by giohappy** [comment]
@imranrajjad and @nmco I don't know if a custom Authentication Filter would be ok here, as it was done for geonode-oauth2 and ApiKey, or if an Authentication Provider is the right way to go.

In any case I fin that:

 - geonode-oauth2 has the most straightforward configuration (and it would be sufficient for my specific purposes): you set the username authentication endpoint and select and ApiKey Role Service to match user roles
![image](https://user-images.githubusercontent.com/571129/74234532-47831100-4ccd-11ea-9ebc-d449d5948702.png)

- ApiKey uses a more convoluted approach: it gives you more functionalities, like mapping remote groups to a subset of local groups (I don't know if anybody uses this), and it uses the optional roles returned by the webservice to match Geoserver groups and/or roles. For roles matching it uses the AuthKey UserGroup Service where the REST Roles Service is selected. It basically decouples the filter and the REST roles service through a UserGroups Service.

**AuthKey Configuration**
![image](https://user-images.githubusercontent.com/571129/74234683-b6606a00-4ccd-11ea-91ac-271532e900e7.png)

**UserGroup Service Configuration**
![image](https://user-images.githubusercontent.com/571129/74234718-c7a97680-4ccd-11ea-8670-106ff72ff906.png)

For our use case having something simpler, similar to geonode-oauth, is fine, but I'll let  @nmco express itself on this.

**Comment by imranrajjad** [comment]
@giohappy thanks for the detailed clarification,

After going through the basic user:password based authentication and how authKey works. Below are my findings and a revised plan of implementation.

Since these requirements do not involve key mechanism, hence this implementation will go in the core and not the authkey extension.

- Currently, as rightly pointed by @aaime , [GeoServerBasicAuthenticationFilter.java#L48](https://github.com/geoserver/geoserver/blob/8b41abb4294e94c8fa10e6793f72a466a70d62e3/src/main/src/main/java/org/geoserver/security/filter/GeoServerBasicAuthenticationFilter.java#L48) is a wrapper around `org.springframework.security.web.authentication.www.BasicAuthenticationEntryPoint` to which credentials based authentication is delegated to and this is the part that this requirement wishes to carry out using an external web service. 
- Since we are concerned with user/pass like credentials. It will make sense to introduce a new **Authorization Provider** which delegates calls to an external web service. The interaction with the external web service can be made generic and configurable by taking inspiration for Authkey, where users can define a an external end point with placeholders. The authentication process can be similar to whats happening in [WebServiceAuthenticationKeyMapper.callWebService](https://github.com/geoserver/geoserver/blob/e893e997434f0ca1b085896e27b1d774f8cafb6a/src/extension/authkey/src/main/java/org/geoserver/security/WebServiceAuthenticationKeyMapper.java#L219).

For processing roles, there can be two scenarios

1. Roles are returned against a successful authentication call
2. The service only returns confirmation in terms of content (which can be parsed using regex) or HTTP code.

To cater for both scenarios, the Provider GUI page can allow user to either configure a local service available in Geoserver or it can allow user to parse response to extract roles, again inspiration can be taken from how its being in authKey. Something similar to [WebServiceBodyResponseUserGroupService.extractRoles](https://github.com/geoserver/geoserver/blob/ad7eb9041ff5dab135cba61798e8415c3b91097d/src/extension/authkey/src/main/java/org/geoserver/security/WebServiceBodyResponseUserGroupService.java#L147) but embeded inside the new Authorization Provider. This will handle both scenarios where a web service responds with authorities and where its just confirmation of credentials.

I would be glad to hear everyone`s thoughts on this.

**Comment by nmco** [comment]
@giohappy and @imranrajjad  I don't understand nothing about GeoServer security authentication neither as a user, and certainly not code wise, so please read my comments taking that into consideration.

That say, I had a quick read through the GeoServer documentation and  based on this description from @giohappy :

> I would like to implement something similar for Basic Auth access. From my understanding Geoserver's Basic Auth filter only extracts username and password credentials from the request, then it delegates the authentication to the Authentication Providers. The available APs are "username and password" (which uses a UserGroups service to perform the authentication), "LDAP" and "JDBC".
If we had a "Web Service Authentication Provider", able to perform the authentication of the user credentials (obtained from the Basic Auth filter) and return the username, I should have (in few workds) "Basic Auth with external authentication service"

It looks to me we have two needs:
- A new authentication provider, to authenticate the user and obtain a _principal_.
- A new roles service, to obtain the authenticated user _authorities_.

Then there is this (insane) security system architecture, which seems to imply that roles can be obtained directly without have the groups service in the middle, both sometimes not.

Long story short we (@nmco, @imranrajjad and @giohappy ) need to have a call tomorrow to discuss this to understand what's actually needed in the user perspective, what can we do to implement this and make generic.

**Comment by giohappy** [comment]
@imranrajjad here below I attach an example from a working configuration of an AuthKey REST Roles Service (I will share the credentials in a private channel of course)

### Geoserver Configuration
![image](https://user-images.githubusercontent.com/571129/74433264-5c43de00-4e60-11ea-9306-7c3367ef7c3b.png)

### Sample requests
**Users endpoint**
(I'll send you <api_key> in a private channel to test it)

Request:
```
curl -X GET \
  https://dev.geonode.geo-solutions.it/api/users \
  -H 'Accept: */*' \
  -H 'Accept-Encoding: gzip, deflate' \
  -H 'Authorization: ApiKey <api_key>' \
  -H 'Cache-Control: no-cache' \
  -H 'Host: dev.geonode.geo-solutions.it' \
```

Response (partial view):
```
{
    "users": [
        {
            "username": "AnonymousUser",
            "groups": [
                "anonymous"
            ]
        },
        {
            "username": "henryhansen",
            "groups": [
                "anonymous",
                "registered-members"
            ]
        }
    ]
}
```

**Roles and adminRole endpoints**
Same as above, just change the endpoint URLs as from the Geoserver configuration.

**Comment by imranrajjad** [comment]
@nmco and @giohappy thanks for the meeting and clarification. Below please find the updated plan after our call yesterday.

### Litmus Test

- Configured a mock Authkey Rest Role service and AuthKEY WebService Body Response UserGroup
- Configured an Authkey filter with web service mapper as shown below

![image](https://user-images.githubusercontent.com/11514936/74510448-a17e1380-4f25-11ea-8624-348b34c18f27.png)

With a correct regex to parse the user name from the response, I was able to use role services associated with all of the user groups and not just the Auth Key user groups.

What this means in practice is that taking a username we can use any one of the available role services to authorize users. In case of Authkey mapper the user name is retrieved by providing the webservice with auth key. In the current scenario we will work with credentials directly instead of an key, hence this forces us to introduce an AuthorizationProvider.


### Implementation Plan

#### Authentication
To understand the implementation plan, we can look at the example of authentication and authorization being done inside [JDBCConnectAuthProvider](https://github.com/geoserver/geoserver/blob/e893e997434f0ca1b085896e27b1d774f8cafb6a/src/security/jdbc/src/main/java/org/geoserver/security/jdbc/JDBCConnectAuthProvider.java#L49). Where the credentials are passed to database for authentication purpose and a successful authentication is followed by Authorization using configured user group service. 

In current scenario we will replace Database with a configurable Web Service. Since we will be working in AuthorizationProvider, we will not require to parse username from the response as it is already available as part of credentials.

#### Authorization

Since we are working with Web service here, we can provide option to directly extract roles from web service response as being done in [WebServiceBodyResponseUserGroupService.extractRoles](https://github.com/geoserver/geoserver/blob/e893e997434f0ca1b085896e27b1d774f8cafb6a/src/extension/authkey/src/main/java/org/geoserver/security/WebServiceBodyResponseUserGroupService.java#L147). User can configure the regex in GUI and directly parse roles without delegating username to available Role services for simplicity. 

For using existing Role services, use of the API calls such as `getSecurityManager().listRoleServices() ` and `getSecurityManager().loadRoleService("role_service_name")`  will allow to obtain a Role service instance which can be used to instantiate a RoleCalculator. Which will be used to obtain authorities /roles.


#### GUI

![image](https://user-images.githubusercontent.com/11514936/74512887-1142cd00-4f2b-11ea-993b-9166ba4a3b21.png)

**Comment by giohappy** [comment]
@imranrajjad thanks for the analysis. I think it is a good plan. I'm not sure if your proposal is to implement both alternatives: extracting of roles from the web response or configuring a roles service.

With regard to this do you confirm that extracting the roles from the web service response is enough? I supposed that a Roles Service was required in any case, and one of the reasons was to instantiate a RoleCalculator.
Can you clarify this point?

**Comment by imranrajjad** [comment]
@giohappy 

1. I propose to implement both. Please forgive my mock of radio buttons.
2. Extracting roles from Web service is indeed enough but since this has to be generic, we would want to consider the case where service does not return roles, in that scenario we will have to use an available role service, for simplicity which could the be the default one as being done here : [WebServiceBodyResponseUserGroupService.extractRoles](https://github.com/geoserver/geoserver/blob/e893e997434f0ca1b085896e27b1d774f8cafb6a/src/extension/authkey/src/main/java/org/geoserver/security/WebServiceBodyResponseUserGroupService.java#L147). So yes when extracting roles from web service response, the default role service will come into effect

**Comment by nmco** [comment]
Thank you for the plan @imranrajjad it matches indeed what we discuss during our meeting :+1:.
 
@giohappy the two options ``Parse roles from web service`` and ``Role service`` are mutually exclusive, for your use case @giohappy I think you only need the first one. That say as @imranrajjad stayed, the second option will make this solution generic enough to go into core, and I suspect that the effort to implement it will be quite small. **We will provide an estimate for both and you can make a final decision @giohappy.**

A few technical questions \ comments I have:
- The user name and passwords will be send in clear to the web authentication service if HTTPS is not used, this shoudl be stressed out in the documentation.
- Since the authentication for each request will imply a remote request to another service some caching should be used in the middle, is there already some generic cache in the security system that can be used @imranrajjad? Or is there some session mechanism that will be used? Can you have look around to what's already done in GeoServer security system a clarify this point?  

I have create [this task sheet for the estimates](https://docs.google.com/spreadsheets/d/1xpbkmLrwRQYZucQELT7gURUbIPuw2VRSi91xDyQ6sWA/edit#gid=376642658), @imranrajjad make sure you create tasks for each independent functionality so @giohappy can pick what matches his budget. Take this into account when write the estimates:
- Since we are targeting core a GSIP will be required, mention explicitly that we are targeting 2.18-RC, 2.17.1 and 2.16.4.
- Documentation in GeoServer will need be updated with clear instruction in how to use this new security mechanism.  
- Integration tests will need to be implemented, this means testing with the GeoServer HTTP mock and caching in between.

**Comment by giohappy** [comment]
sorry for the insistence but let me see if I understood it correctly.

 - in case the (alternative) option to extract users from the service response, the default Roles service will be used under the hood
 - otherwise a Roles service can be configured. In this case the webservice doesn't need to return the roles inside the response, because the mapping between user and role will be done by the Roles service.

It's not clear to me why the latter case (which might also be configured to use the default Roles service) doesn't need the roles from the webservice response, while the first requires them (and still uses the default Roles service).

**Comment by imranrajjad** [comment]
@giohappy no need to be sorry 😄 I am will try to clarify in easy words.

1. Option 1 : When using a configured Location service, after the web service authenticates the user, the user`s roles are extracted from the configured role service (can be default or anyother). clear?
. 
2. Option 2 : Web service responds with roles, we extract the roles using regex and treat them as authorities. Then following the pattern of [WebServiceBodyResponseUserGroupService.extractRoles](https://github.com/geoserver/geoserver/blob/e893e997434f0ca1b085896e27b1d774f8cafb6a/src/extension/authkey/src/main/java/org/geoserver/security/WebServiceBodyResponseUserGroupService.java#L147), we will also peek in the default system role service and merge any roles in existing set if found against the username 😃 

Through GUI user can opt for any of these workflows.

**Comment by giohappy** [comment]
Ok, for the second case, peeling the roles from the default roles service is not strictly required. Obatining the roles from the service response would be enough, right?

It seems redundant to me doing the extra step of "merging" roles both from regex extraction and default roles service. But if you think it's the right way to go I'm ok ;)

**Comment by nmco** [comment]
@giohappy exact, for you use case you only need option 1, option 2 will be there just to make this more generic.

**Comment by nmco** [comment]
I ha d a quick chat whit @giohappy, in option 1 when an user authenticates we can obtain the list of roles associated whit the user. But, we actually need roles to configure the data security rules. How does that works for AuthKey, are roles create on GeoServer? or there is a call to the configured roles service to obtain the list of availble roles?

**Comment by imranrajjad** [comment]
In authkey after we get the username against the passed key, the username is passed to attached user/group service which delegates it to a either a roles service. If authkey mapper is configured to use a web service, the response is passed to authkey web response group and roles are extracted from the initial response of web service.

The data security is not coupled with user assigned roles. They come into effect when user tries to access the data.

**Comment by imranrajjad** [comment]
@nmco 

A generic caching mechanism for basic authentication is already available which will be easily res-used in our case. A successful call of [GeoServerAuthenticationProvider.authenticate](https://github.com/geoserver/geoserver/blob/master/src/main/src/main/java/org/geoserver/security/GeoServerAuthenticationProvider.java#L52), ends up being persisted inside the session as a  Basic Authentication token along with relevant request attributes which work as hints/signals for filters inside FilterChain to differentiate between an authenticated, new or invalid session.

This was validated by debugging the `UsernamePasswordAuthenticationProvider`, which is only called once for the first successful login. Afterwards the credentials are converted into authentication token which is accessed by  [GeoServerBasicAuthenticationFilter.getCacheKey](https://github.com/geoserver/geoserver/blob/master/src/main/src/main/java/org/geoserver/security/filter/GeoServerBasicAuthenticationFilter.java#L86) to instantiate `Authentication `, for re-usability.

For the authorities, they are cached at [GeoServerSecurityManager.getAuthenticationCache](https://github.com/geoserver/geoserver/blob/master/src/main/src/main/java/org/geoserver/security/GeoServerSecurityManager.java#L689)

**Comment by nmco** [comment]
@imranrajjad as we discussed i our quick call, double check where the roles that will be available to configure that data rules come from:
![image](https://user-images.githubusercontent.com/1590542/74662952-d2af4b80-5192-11ea-8a31-99c63ed5728f.png)

**Comment by imranrajjad** [comment]
@nmco 

Given a Geoserver with three Role services, with 1 default and 2 REST services as below

![image](https://user-images.githubusercontent.com/11514936/74710219-ec0fd080-5242-11ea-907f-36eedbba9528.png)

The Data Security page shows up **ALL** available roles.

![image](https://user-images.githubusercontent.com/11514936/74710310-1e213280-5243-11ea-8af6-e7648cd695f2.png)

**Comment by nmco** [comment]
Thank you for the feedback @imranrajjad, So if I understood correctly in practical terms:
- We need a new authentication provider, that will perform an authentication against a remote web service
- A role service, that could be the AuthKey one for example, where this role service needs to be able to at least:
  - Retrieve the users for a certain user
  - Retrieve all the available roles

Configuring the system will require:
- Make sure that a role service or user\group service is configured
- Configure the new authentication provider:
  - Selecting either a user\group or role service directly

Did I understood it correctly @imranrajjad ? Does this estimate still stand?
https://docs.google.com/spreadsheets/d/1xpbkmLrwRQYZucQELT7gURUbIPuw2VRSi91xDyQ6sWA/edit#gid=376642658

**Comment by imranrajjad** [comment]
@nmco

we don't need a new role service, we will make use of any of the existing role services.
In case user decides to extract roles from the authentication response, we will still peek in system default role service.

Indeed the estimate stands, the unit testing will need to extensive.

**Comment by giohappy** [comment]
@imranrajjad what about adding the option to send username/password inside a custom header field, instead of GET params?
My concern is that GET params will be stored inside logs.

My idea is the following: add an checkbox to define a header name (e.g. "X-HTTP-AUTHORIZATION") and pass username:password using the standard base64 encoding

If the checkbox is not flagged username and password will be sent as GET params ("username" and "password").

Could you share the plugin as soon as it is in a testable shape? So that we can do a quick test it works fine in our context.

**Comment by nmco** [comment]
Starting from the button @giohappy:

> Could you share the plugin as soon as it is in a testable shape? So that we can do a quick test it works fine in our context.

Yes sure.

> @imranrajjad what about adding the option to send username/password inside a custom header field, instead of GET params?
My concern is that GET params will be stored inside logs.

My idea is the following: add an checkbox to define a header name (e.g. "X-HTTP-AUTHORIZATION") and pass username:password using the standard base64 encoding

If the checkbox is not flagged username and password will be sent as GET params ("username" and "password").

This depends on the end-point you have, if you want to authenticate using an header? Fine. Using GET parameters? Fine. In both cases it may end up on the log of GeoServer or logged by any other component of the network, for example a proxy.

 If you want to make this secure HTTPS shoudl be used, otherwise we will need to engenier something.

**Comment by imranrajjad** [comment]
@giohappy I will let you know, we will be using the `org.geotools.data.ows.HTTPClient`, should be possible

**Comment by giohappy** [comment]
@nmco I know that headers can be logged the same as GET parameters, but in general there are far more ways to leak GET params (referrer headers, logged URLs).
Of course in a m2m scenario the risk is reduced, since we don't have browser history, etc., so my proposal is not critical here. 

Anyway If adopting headers sounds an overhead without real benefits to you, or it would involve a significant effort, just forget it ;)

**Comment by nmco** [comment]
Doesn't look significant to me, but let's wait for @imranrajjad feedback.

**Comment by giohappy** [comment]
On second thought I also agree that it isn't that significant, although logging practices expose URL params more then headers.
We can forget this option for the moment.

**Comment by nmco** [comment]
Just for the record, we are moving on whit this (@imranrajjad already started to work on the implementation), including the HTTP header option (you have a good point here @giohappy).

@giohappy we will share whit you a testable version as soon as possible, once you got your feedback and are all happy will create a GSIP and move on whit the contribution.

**Comment by imranrajjad** [comment]
@nmco @giohappy 

Some snapshots of GUI

#### When reading roles from service response
![image](https://user-images.githubusercontent.com/11514936/75425524-b8c8f200-5964-11ea-8008-986107e866d7.png)

#### When using a local role service
![image](https://user-images.githubusercontent.com/11514936/75425590-dbf3a180-5964-11ea-8ba1-4f3ed83564eb.png)

#### Download Test WAR
A test  WAR file is available for [DOWNLOAD](https://drive.google.com/open?id=1LZiUTR6oowkekq9YBYPMUH0ae2bwnLqq)


Looking forward to your feedbacks

**Comment by giohappy** [comment]
thanks @imranrajjad we will test it ASAP

**Comment by nmco** [comment]
On which branch is the code @imranrajjad ?

**Comment by nmco** [comment]
Answer from @imranrajjad:
https://github.com/imranrajjad/geoserver/tree/issue_145_auth

**Comment by imranrajjad** [comment]
Working branch :
 https://github.com/geosolutions-it/geoserver/tree/issue_145_auth

**Comment by nmco** [comment]
@giohappy we will wait for your feedback before implement the remaining test and submit the GSIP :small_airplane:.

**Comment by giohappy** [comment]
yes @nmco I think I will be able to give you a feedback on Monday.

**Comment by nmco** [comment]
That's great @giohappy, thank you!

**Comment by giohappy** [comment]
@imranrajjad I've deployed the WAR but I get the following kind of errors when trying to access Layer Preview, Tile Layers settings, Layer Groups, etc.

NOTICE: I haven't configured the new auth provider yet.

```
2020-03-02 15:53:55,719 DEBUG [org.geoserver.security.IncludeQueryStringAntPathRequestMatcher] - Checking match of request : 'Path: /web/, QueryString: null'; against '/web/**'
2020-03-02 15:53:55,722 DEBUG [org.geoserver.security.IncludeQueryStringAntPathRequestMatcher] - Matched Path: /web/, QueryString: null with /web/**
2020-03-02 15:53:55,823 DEBUG [org.geoserver.filters] - Creating a new http session inside the web UI (normal behavior)
java.lang.Exception: Full stack trace for the session creation path
	at org.geoserver.filters.SessionDebugFilter$SessionDebugWrapper.getSession(SessionDebugFilter.java:92)
	at javax.servlet.http.HttpServletRequestWrapper.getSession(HttpServletRequestWrapper.java:231)
	at javax.servlet.http.HttpServletRequestWrapper.getSession(HttpServletRequestWrapper.java:231)
	at javax.servlet.http.HttpServletRequestWrapper.getSession(HttpServletRequestWrapper.java:231)
	at org.apache.wicket.session.HttpSessionStore.getHttpSession(HttpSessionStore.java:95)
	at org.apache.wicket.session.HttpSessionStore.getSessionId(HttpSessionStore.java:166)
	at org.apache.wicket.Session.bind(Session.java:267)
	at org.apache.wicket.Session.replaceSession(Session.java:520)
	at org.geoserver.web.GeoServerApplication.onApplicationEvent(GeoServerApplication.java:439)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.doInvokeListener(SimpleApplicationEventMulticaster.java:172)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.invokeListener(SimpleApplicationEventMulticaster.java:165)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:139)
	at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:402)
	at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:359)
	at org.springframework.security.authentication.DefaultAuthenticationEventPublisher.publishAuthenticationSuccess(DefaultAuthenticationEventPublisher.java:99)
	at org.springframework.security.authentication.ProviderManager.authenticate(ProviderManager.java:224)
	at org.geoserver.security.GeoServerSecurityManager.checkForDefaultAdminPassword(GeoServerSecurityManager.java:1374)
	at org.geoserver.security.web.SecurityHomePageContentProvider$SecurityWarningsPanel.<init>(SecurityHomePageContentProvider.java:116)
	at org.geoserver.security.web.SecurityHomePageContentProvider.getPageBodyComponent(SecurityHomePageContentProvider.java:44)
	at org.geoserver.web.GeoServerHomePage$1.populateItem(GeoServerHomePage.java:133)
	at org.apache.wicket.markup.html.list.ListView.onPopulate(ListView.java:523)
	at org.apache.wicket.markup.repeater.AbstractRepeater.onBeforeRender(AbstractRepeater.java:124)
	at org.apache.wicket.Component.internalBeforeRender(Component.java:950)
	at org.apache.wicket.Component.beforeRender(Component.java:1018)
	at org.apache.wicket.MarkupContainer.onBeforeRenderChildren(MarkupContainer.java:1836)
	at org.apache.wicket.Component.onBeforeRender(Component.java:3916)
	at org.apache.wicket.Page.onBeforeRender(Page.java:801)
	at org.apache.wicket.Component.internalBeforeRender(Component.java:950)
	at org.apache.wicket.Component.beforeRender(Component.java:1018)
	at org.apache.wicket.Component.internalPrepareForRender(Component.java:2236)
	at org.apache.wicket.Page.internalPrepareForRender(Page.java:242)
	at org.apache.wicket.Component.render(Component.java:2325)
	at org.apache.wicket.Page.renderPage(Page.java:1018)
	at org.apache.wicket.request.handler.render.WebPageRenderer.renderPage(WebPageRenderer.java:124)
	at org.apache.wicket.request.handler.render.WebPageRenderer.respond(WebPageRenderer.java:236)
	at org.apache.wicket.core.request.handler.RenderPageRequestHandler.respond(RenderPageRequestHandler.java:175)
	at org.apache.wicket.request.cycle.RequestCycle$HandlerExecutor.respond(RequestCycle.java:895)
	at org.apache.wicket.request.RequestHandlerStack.execute(RequestHandlerStack.java:64)
	at org.apache.wicket.request.cycle.RequestCycle.execute(RequestCycle.java:265)
	at org.apache.wicket.request.cycle.RequestCycle.processRequest(RequestCycle.java:222)
	at org.apache.wicket.request.cycle.RequestCycle.processRequestAndDetach(RequestCycle.java:293)
	at org.apache.wicket.protocol.http.WicketFilter.processRequestCycle(WicketFilter.java:261)
	at org.apache.wicket.protocol.http.WicketFilter.processRequest(WicketFilter.java:203)
	at org.apache.wicket.protocol.http.WicketServlet.doGet(WicketServlet.java:137)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:635)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:742)
	at org.springframework.web.servlet.mvc.ServletWrappingController.handleRequestInternal(ServletWrappingController.java:166)
	at org.springframework.web.servlet.mvc.AbstractController.handleRequest(AbstractController.java:177)
	at org.springframework.web.servlet.mvc.SimpleControllerHandlerAdapter.handle(SimpleControllerHandlerAdapter.java:52)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1040)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:943)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:898)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:635)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:742)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.geoserver.filters.ThreadLocalsCleanupFilter.doFilter(ThreadLocalsCleanupFilter.java:26)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.geoserver.filters.SpringDelegatingFilter$Chain.doFilter(SpringDelegatingFilter.java:69)
	at org.geoserver.wms.animate.AnimatorFilter.doFilter(AnimatorFilter.java:73)
	at org.geoserver.filters.SpringDelegatingFilter$Chain.doFilter(SpringDelegatingFilter.java:66)
	at org.geoserver.filters.SpringDelegatingFilter.doFilter(SpringDelegatingFilter.java:41)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.geoserver.platform.AdvancedDispatchFilter.doFilter(AdvancedDispatchFilter.java:37)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:320)
	at org.geoserver.security.filter.GeoServerCompositeFilter$NestedFilterChain.doFilter(GeoServerCompositeFilter.java:70)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:127)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:91)
	at org.geoserver.security.filter.GeoServerCompositeFilter$NestedFilterChain.doFilter(GeoServerCompositeFilter.java:74)
	at org.geoserver.security.filter.GeoServerCompositeFilter.doFilter(GeoServerCompositeFilter.java:91)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.geoserver.security.filter.GeoServerCompositeFilter$NestedFilterChain.doFilter(GeoServerCompositeFilter.java:70)
	at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:119)
	at org.geoserver.security.filter.GeoServerCompositeFilter$NestedFilterChain.doFilter(GeoServerCompositeFilter.java:74)
	at org.geoserver.security.filter.GeoServerCompositeFilter.doFilter(GeoServerCompositeFilter.java:91)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.geoserver.security.filter.GeoServerAnonymousAuthenticationFilter.doFilter(GeoServerAnonymousAuthenticationFilter.java:51)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.geoserver.security.filter.GeoServerCompositeFilter$NestedFilterChain.doFilter(GeoServerCompositeFilter.java:70)
	at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:200)
	at org.geoserver.security.filter.GeoServerCompositeFilter$NestedFilterChain.doFilter(GeoServerCompositeFilter.java:74)
	at org.geoserver.security.filter.GeoServerCompositeFilter.doFilter(GeoServerCompositeFilter.java:91)
	at org.geoserver.security.filter.GeoServerUserNamePasswordAuthenticationFilter.doFilter(GeoServerUserNamePasswordAuthenticationFilter.java:122)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.geoserver.security.filter.GeoServerCompositeFilter$NestedFilterChain.doFilter(GeoServerCompositeFilter.java:70)
	at org.springframework.security.web.authentication.rememberme.RememberMeAuthenticationFilter.doFilter(RememberMeAuthenticationFilter.java:158)
	at org.geoserver.security.filter.GeoServerCompositeFilter$NestedFilterChain.doFilter(GeoServerCompositeFilter.java:74)
	at org.geoserver.security.filter.GeoServerCompositeFilter.doFilter(GeoServerCompositeFilter.java:91)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.geoserver.security.filter.GeoServerCompositeFilter$NestedFilterChain.doFilter(GeoServerCompositeFilter.java:70)
	at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:105)
	at org.geoserver.security.filter.GeoServerSecurityContextPersistenceFilter$1.doFilter(GeoServerSecurityContextPersistenceFilter.java:52)
	at org.geoserver.security.filter.GeoServerCompositeFilter$NestedFilterChain.doFilter(GeoServerCompositeFilter.java:74)
	at org.geoserver.security.filter.GeoServerCompositeFilter.doFilter(GeoServerCompositeFilter.java:91)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:215)
	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:178)
	at org.geoserver.security.GeoServerSecurityFilterChainProxy.doFilter(GeoServerSecurityFilterChainProxy.java:142)
	at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:358)
	at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:271)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.geoserver.filters.LoggingFilter.doFilter(LoggingFilter.java:101)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.geoserver.filters.XFrameOptionsFilter.doFilter(XFrameOptionsFilter.java:79)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.geoserver.filters.GZIPFilter.doFilter(GZIPFilter.java:47)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.geoserver.filters.SessionDebugFilter.doFilter(SessionDebugFilter.java:46)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.geoserver.filters.FlushSafeFilter.doFilter(FlushSafeFilter.java:42)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:200)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:493)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:137)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:81)
	at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:660)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:87)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:798)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:806)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1498)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by imranrajjad** [comment]
@giohappy the error logs look un-related however I have validated that they do occur on my Tomcat 8.5 instance.

I have re-based the work on fresh master branch and validated that GUI errors do not occur anymore.

Please [DOWNLOAD](https://drive.google.com/file/d/1i6EDRlBh46BdgL0JCWfTAYIcL_udIfjE/view?usp=sharing) from here

Awaiting your feedback

**Comment by giohappy** [comment]
@imranrajjad to have consistency between auth web services the groups/roles parsed from the webservice body response should be transformed into the form ROLE_<ROLENAME>, where ROLENAME is the capitalized role parsed from the response.

E.g. 

my_user_role1 -> ROLE_MY_USER_ROLE1
another_custom_user_role -> ROLE_ANOTHER_CUSTOM_USER_ROLE
External_Role_X -> ROLE_EXTERNAL_ROLE_X

This is what both AuthKEY WebService Body Response UserGroup Service and AuthKey REST Role Service do.

So, as a resume:

1 - roles parsed from response: the autnetication provider must do the mapping (capitalize + prefix)
2 - when using a REST Role Service the same mapping is already done by the service

**Comment by giohappy** [comment]
@imranrajjad as agreed in the call I share with you an example response from my webservice and the regex I set to parse the roles.
It seems to be correct but the Basic Auth Web Service doesn't seem to assing the right authoritites to the user (always Forbidden response).

Regex:
`^.*?"groups"\s*:\s*\["([^"]+)"\].*$`

Response example:
`{"username":"giovanni.allegri@geo-solutions.it","groups":["hub,free"]}`

**Comment by imranrajjad** [comment]
@giohappy as discussed, I was unable to re-produce the issue after spending sometime after configuring the WAR file with similar security configurations.

**Comment by giohappy** [comment]
I'm going to schedule a troubleshotting session with you

**Comment by nmco** [comment]
Me and @imranrajjad performed a debug session with sharing screen using the provided data directory. Long story short, turns out that  the tests were performed using WMS service, but a service rule restraining the WMS operations to administrators only was configured:
```
#Fri Mar 06 15:09:34 CET 2020
wfs.DropStoredQuery=GROUP_ADMIN,ADMIN
wms.*=GROUP_ADMIN,ADMIN
wfs.CreateStoredQuery=GROUP_ADMIN,ADMIN
```
This was the cause for the ``403 Forbidden``, we performed a few (quick) tests playing a bit whit the data rules and security rules and looks like things are working as expected.

@giohappy this should unblock you on your tests, that say looks like we can move on and make the GSIP proposal public and start the public discussion, what do you think @giohappy ?

A side note, its not easy to understand what's going on with in terms of security in GeoServer by only looking at the log, this is something we need to think about in future iterations whit the security sub-system.

**Comment by nmco** [comment]
@imranrajjad were are we with the GSIP proposal for this?

**Comment by nmco** [comment]
Summarizing the results of my tests and discussion with @giohappy and @imranrajjad :
1) The system fails to handle the situation where multiple roles are provided, 403 will be returned even if the expected role is there (the order doesn't matter), this is the cause of the error @giohappy  is experiencing, for example:
```json
{
    "groups": [
        "free",
        "NOBODY"
    ],
    "username": "NUNO"
}
```
2) @imranrajjad did ha da look at this feedback from @giohappy https://github.com/geosolutions-it/geoserver/issues/145#issuecomment-593866491, can you dig this a bit?

3) Currently looks like we have no control on GeoServer authorization system cache (this is different from the session), its even not clear which parameters are used to instantiated or what's cached, I create issue for this discussion https://github.com/geosolutions-it/geoserver/issues/162.

@imranrajjad, top priority is fixing 1) then 2).

**Comment by imranrajjad** [comment]
@giohappy
can you please check your regex expression. I have used this website to work the regex expression (https://www.freeformatter.com/java-regex-tester.html#ad-output)

Using the regex expression : `^.*?"groups"\s*:\s*\["([^"]+)"\].*$`

When evaluating the web response with group names as array elements `{"username":"imran.rajjad@geo-solutions.it","groups":["free","hub"]}`, nothing is extracted

However, if we make both groups part of single string inside array like `{"username":"imran.rajjad@geo-solutions.it","groups":["free,hub"]}`, we are able to extract roles and there is no 403 error. 

Can you please verify this on your end?

thanks

**Comment by giohappy** [comment]
Ok,  reviewing it my case is not the same as the one reported by @nmco because my service returns a single string:

`{"username":"giovanni.allegri@geo-solutions.it","groups":["free,hub"]}`

but still it doesn't work for me.

**Comment by giohappy** [comment]
@imranrajjad it still doesn't work for me, even with:

- the regex you proposed me to test on skype
- single group returned for test user

please can you share your WAR with the configurations ready to test?

**Comment by nmco** [comment]
@giohappy and @imranrajjad I re-based the development branch with master:
https://github.com/geosolutions-it/geoserver/tree/issue_145_auth

I built I clean WAR, and used a clean GeoServer release data directory to setup everything from scratch, and it seems to work.

You can find the GeoServer WAR, data directory and Flask HTTP static server I used here:
https://www.dropbox.com/work/Backups/MAPSTAND/Basic%20Auth%20Porvider%20%23145/19-03-2020

@giohappy feel free to ping me for support on setting this up on you environment.

**Comment by imranrajjad** [comment]
@all

The GSIP is available for review at : https://github.com/geoserver/geoserver/wiki/GSIP-190---Web-Service-Authentication

**Comment by aaime** [comment]
Seems a new authentication plugin doing its own (unusual, at least has not been required for years) thing... should be a community module, not a change to core. So no need for a GSIP, just a mail to ask for the community module slot?

**Comment by giohappy** [comment]
@nmco I was just setting up the environment to test it again :)
I'll let you know...

**Comment by giohappy** [comment]
@imranrajjad and @nmco I've spot the issue with my service. By default its response wasn't JSON formatted. My fault.

@imranrajjad please proceed with the transformation of role names to ROLE_capital(role), following the pattern assumed by AuthKey.

**Comment by giohappy** [comment]
@imranrajjad as reported in our chats tests went fine, so from my side nothing to add. You can proceed with the packaging of the community module.

**Comment by giohappy** [comment]
@imranrajjad @nmco where we are with this? I can't find if it has been merged, backported, etc.

**Comment by nmco** [comment]
If my memory serves me well @giohappy, the client didn't need this any-more hence the priority become very low, this was not merged yet, right @imranrajjad ?

**Comment by giohappy** [comment]
does it take much work to merge it? I don't want to leave this pending, since we'll need it soon.

**Comment by imranrajjad** [comment]
@aaime directed to make this available as a community module instead of making it part of core. However, the PR for community was not submitted as the unit tests still need to be migrated to community module.

**Comment by giohappy** [comment]
how mich work is left to complete this?

**Comment by imranrajjad** [comment]
@giohappy apologies, I somehow missed this,

At least one day worth of work is left which includes:

- Notifying the community about the new community module
- Migration of unit tests
- Validating and testing community module assembly

We also need to follow the community guidelines on  [community-modules guidelines](https://docs.geoserver.org/latest/en/developer/policies/community-modules.html#process)

**Comment by giohappy** [comment]
@simboss even if we have reduced the dev time for MapStand I would complete this part, otherwise it gets lost and we've wated resources. What do you think?

**Comment by simboss** [comment]
@giohappy I agree

@nmco can you make sure this work gets completed?
I would need to do things like this was geosevrer core and not a community module:
 
- complete documentation 
- code coverage as high as possible
- add it to the build

**Comment by nmco** [comment]
Understood @simboss, @imranrajjad as a few urgent tasks to to finish for EMSA, end of next week we should be able to submit a PR by the end of next week.

@imranrajjad, things that should included:
- A community module, that's included in the community build and on the GeoServer WAR when the associated profile is used.
- A ZIP should be generated for the community module during the build, like for all the others community modules.
- A GeoServer documentation page should be added for this plug-in [here](https://docs.geoserver.org/stable/en/user/community/index.html).
- In this case I would say at least an 80% test coverage, make sure you add some corner cases to the test cases and the scenario that @giohappy will need.

@imranrajjad please send a mail ASAP to the GeoServer DEV list asking permission to create this community module.

**Comment by imranrajjad** [comment]
@nmco understood, will do.

**Comment by simboss** [comment]
@imranrajjad @nmco where are we on this one?

**Comment by imranrajjad** [comment]
@simboss unfortunately I am busy with closing some critical EMSA deliverables. This is one of my highest priorities. My apologies for the delay.

**Comment by imranrajjad** [comment]
draft PR for community extension : https://github.com/geoserver/geoserver/pull/4483

**Comment by simboss** [comment]
@aaime can you review on mon?

**Comment by afabiani** [comment]
put some comments to the PR [here](https://github.com/geoserver/geoserver/pull/4483#pullrequestreview-517478569)

**Comment by taba90** [comment]
I made a new pull request with a branch that is a checkout from the one in the previous pr, and added some commits.
https://github.com/geoserver/geoserver/pull/4906
@afabiani  and @aaime for the review, one note about [this piece of code](https://github.com/geoserver/geoserver/blob/d90b495a0ab900cbc0a82a7ba117e44025926512/src/community/authbasic/src/main/java/org/geoserver/security/auth/web/SimpleWebServiceAuthenticationProvider.java#L194) that adds a ROLE_ prefix to the roles retrieved from the external services: it seems done with full knowledge of the facts, but seems to me that will imply also that [will never allow to detect and admin from the response roles](https://github.com/geoserver/geoserver/blob/d90b495a0ab900cbc0a82a7ba117e44025926512/src/community/authbasic/src/main/java/org/geoserver/security/auth/web/SimpleWebServiceAuthenticationProvider.java#L243) if the Geoserver admin role doesn't start with "ROLE_". For the moment I left it as it was.

**Comment by giohappy** [comment]
@taba90 if I remember correctly the role services retrieve the actual admin role from a specific endpoint (see image below). This endpoint returns the role to be considered a Geoserver admin, independently from the ROLE_* convention.

The convention of the ROLE_* prefix stems from the AuthKey module. It's exaplined inside [AuthKEY WebService Body Response UserGroup Service](https://docs.geoserver.org/stable/en/user/community/authkey/index.html#authkey-webservice-body-response-usergroup-service) ("Additional Options" sub-paragraph)

![image](https://user-images.githubusercontent.com/571129/112345387-45a7c180-8cc5-11eb-986b-14f65f34b345.png)

**Comment by taba90** [comment]
thanks for the heads up @giohappy  now its clear. Currently there is no possibility to specify a separate regex/path for the admin role   in the new community module. 
Let me know if this should be implemented. For the moment I've simply limited to make the module ready to be contributed with the functionality already in place.

**Comment by nmco** [comment]
@taba90 for the moment let's focus on the functionalities already in place (this module if I'm not mistaken was already tested by @giohappy). @giohappy if you would like that extra functionality I would recommend to create a new issue.

**Comment by giohappy** [comment]
I haven't asked anything :)
A long time has passed since the last tests. Do you want to test it again @taba90 before merging?

**Comment by taba90** [comment]
the old pr is still open.  If it is meant to be closed can some one do it? (I don't have permissions) https://github.com/geoserver/geoserver/pull/4483

**Comment by taba90** [comment]
Attaching here the 2.18.2 jar for test purpose
[gs-authbasic-2.18.2.zip](https://github.com/geosolutions-it/geoserver/files/6211050/gs-authbasic-2.18.2.zip)

**Comment by simboss** [comment]
@giohappy please, liaise with @taba90 and maybe michal for testing this before we merge it.

**Comment by giohappy** [comment]
@simboss test were already done on Friday. Michal confirms that it's working.

**Comment by simboss** [comment]
perfect.

@taba90 I have not checked the PR but I would want to see clear documentation provided as part of this PR.

If we do not have it, please, open a seprate JIRA and add it in a separate PR.
Make also sure code coverage is enough.

**Comment by simboss** [comment]
K, I see there is basic documentation but I am not sure it suffices.

We need _user_ documentation so that someone who wants to use this feature does not have to become crazy to understand how it works.

**Comment by aaime** [comment]
I've reviewed the module, there are two significant issue.

Naming
----------

First one is naming: it should not be called basic auth, it's confusing (I see this has been discussed above, but it's worth repeating).

The security subsystem is complicated enough already, without adding extra confusion to existing terminology.

The basic authentication is one of the ways a username/password can be provided to a server, and it's already managed by a core functionality, in the form of an authentication filter. 

What the module contains is an authentication provider, which takes an already populated Authentication object (populated by an auth filter) and calls onto an external service to validate the username/password. This is not different from the LDAP authentication or the JDBC connection authentications. See for reference [the GeoSolutions training](https://docs.geoserver.geo-solutions.it/edu/en/security/authentication_providers.html#:~:text=Authentication%20Providers%20are%20pluggable%20components,in%20the%20standard%20GeoServer%20installation) and [the GeoServer own documentation](https://docs.geoserver.org/stable/en/user/security/auth/providers.html)

Since the module documentation and the UI labels call it "Web Service authentication", I would call it that way, it's less confusing than "basic auth".

Less secure than basic auth itself
-------------------------------------------

Basic authentication is often critisized by being weak, but at least, it's base64 encoding the username and password to avoid accidental credential discovery by admins monitoring the network and the requests performed on it.

This module instead sticks clear text usename and passwords straigth into the URL or in the headers, with no encoding whatsoever. This is quite bad, the people monitoring the network and services can be trusted, but having usernames and password appear as-is on their screen (where every by-standers can also see them) is unacceptable. Even if the admins were locked in a room alone, they are not necessarily the same people that are managing the user credentials, they should never be presented with them in such a casual way.

Without asking to apply any actual security to it (e.g., encryption) let's at least be as good as basic auth, and base64 encode them? Please?

HTTPS check is too weak
----------------------------------

The code logs a warning if HTTP is used instad of HTTPS. Better than nothing, but the admin is playing ona  GUI when that happens, not looking at the logs, so they won't see the warning. I recommend having an explicit GUI checkbox to allow HTTP connection usage (handy for quick testing I reckon), that an admin has to manually check in order to save a HTTP link.
Mind, a network might be private, that does not make it secure though, if an intruder sniffin HTTP traffic gets to see the urls with credential going around it's gonna be "jackpot" for them.

**Comment by giohappy** [comment]
@aaime I totally agree with base64. This actually was required by myself long time ago [in this comment](https://github.com/geosolutions-it/geoserver/issues/145#issuecomment-587483995)

**Comment by taba90** [comment]
pr updated

**Comment by taba90** [comment]
@giohappy the module has been merged on master.
I've applied the fix requested by Andrea:

- the credentials are sent Base64 encoded.
- there is a new option from UI to enable authentication using an http service.
- added a check on the response code that has to be 200 to be authenticated. Before the checks was only on the response body to be not null.
- module renamed.

I've already done some manual tests, but if before doing any backport you prefer to give a quick test another time in your environment, here is a 2.18.2 version of the plug-in.
[geoserver-2.18.2-web-service-auth.zip](https://github.com/geosolutions-it/geoserver/files/6272990/geoserver-2.18.2-web-service-auth.zip)
Let me know if/when I can go ahead and do the backports.

**Comment by giohappy** [comment]
I will tell Michal inside the chat.

**Comment by simboss** [comment]
@taba90, if you do not hear from @giohappy shortly, move on with the backport up to 2.18

**Comment by nmco** [comment]
@taba please backport.

**Comment by simboss** [comment]
@taba90 we need to go as back as 2.18

Notify @giohappy once done.

**Comment by nmco** [comment]
Back-ports merged, this will be available on the nightly builds starting tomorrow.

**Comment by simboss** [comment]
cool, thx.

Let's leave it open until @giohappy decides to start using it.

**Comment by giohappy** [comment]
@simboss I've opened another issue under support, because I'm not sure if it's something with the Basic Auth module or what else but things are far from working on dev and staging server.

https://github.com/geosolutions-it/support/issues/1419

**Comment by simboss** [comment]
@nmco as per our discussion, please, ask @zvrablik to help with this

**Comment by nmco** [comment]
@simboss there is another issue created by @giohappy for the problem in question:
https://github.com/geosolutions-it/support/issues/1419

Zdenek is already working on it.

@giohappy I think this one should be closed.
